panther_analysis_tool/main.py

```diff
@@ -1106,124 +1106,104 @@ def _run_tests(
-    output = getattr(detection_result, f"{function_name}_output")
-    is_defined = getattr(detection_result, f"{function_name}_defined")
-    failed = function_error is not None
 
+def _check_destinations(test_result: TestResult, destination_by_name: Dict[str, Any]) -> TestResult:
     # For backwards compatibility we can accept invalid destination names,
     # as long as a string is returned. Strict check is enabled when users
     # pass destination names explicitly through command-line parameters.
-    if function_name == "destinations" and failed and not strict_check:
-        # Otherwise we fall back to a best-effort check of the return type only
-        if isinstance(function_error, UnknownDestinationError):
-            exc: UnknownDestinationError = function_error
-            failed = not _check_destinations_type(exc.result())
-
-    # The function has not been executed
-    if not is_defined:
-        display = None
-    else:
-        display = function_error if failed else output
-
-    return {
-        "error": function_error,
-        "output": output,
-        "failed": failed,
-        "has_invalid_type": isinstance(function_error, FunctionReturnTypeError),
-        "display": display,
-    }
-
+    if (
+        test_result.functions.destinationsFunction
+        and test_result.functions.destinationsFunction.error
+        and not bool(destination_by_name)
+    ):
 
-def _check_destinations_type(obj: Any) -> bool:
-    """Checks that the return value of the `destinations` function is a list of strings"""
-    if obj is None:
-        return True
+        error_message = test_result.functions.destinationsFunction.error.message or ""
+        if "UnknownDestinationError" in error_message:
+            # reset the output and error
+            test_result.functions.destinationsFunction.output = error_message.split(" ")[-1]
+            test_result.functions.destinationsFunction.error = None
+            test_result.passed = True
+            # reset test_result as necessary
+            for function_name in vars(test_result.functions):
+                function_result = vars(test_result.functions)[function_name]
+                if function_result:
+                    if function_result.error:
+                        test_result.passed = False
+                        break
+    return test_result
 
-    if not isinstance(obj, list):
-        return False
 
-    return all(isinstance(destination_name, str) for destination_name in obj)
+def _print_test_result(
+    detection: Detection, test_result: TestResult, failed_tests: DefaultDict[str, list]
+) -> None:
+    status_pass = "PASS"  # nosec
+    status_fail = "FAIL"
+    if test_result.passed:
+        outcome = status_pass
+    else:
+        outcome = status_fail
+    # print overall status for this test
+    print("\t[{}] {}".format(outcome, test_result.name))
+
+    # print function output and status as necessary
+    for function_name in vars(test_result.functions):
```

Comment:
Since `functions` is a dataclass you can also use [asdict](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) I think. `vars` is a bit obscure I believe.